### PR TITLE
Add error property to action

### PR DESCRIPTION
**Pull request by sjsyrek**
Thank you for your contribution! :thumbsup:

Please makes sure that these checkboxes are checked before submitting your PR, thank you!

Related issues: https://github.com/piotrwitek/typesafe-actions/issues/116

* [x] PR have a corresponding issue which is approved and it's linked above
* [x] Rebase before creating a PR to keep commit history clean
* [x] Clear node_modules and reinstall all the dependencies: `npm run reinstall`
* [x] Run ci-check npm script: `npm run ci-check`

For bugfixes:
 * [x] Add at least one unit test to cover the bug that have been fixed

For new feature:
 * [ ] Update API docs and tutorial
 * [ ] Add examples to demonstrate new feature
 * [x] Add type unit tests
 * [x] Add runtime unit tests

### CODE REVIEW

**Comment by sjsyrek** [comment]
@piotrwitek First draft. I was trying to order the function overloads correctly, but it has been non-obvious to me what counts as most specific, etc. If it's not obvious, I'm rather new to TypeScript and in the process of migrating a code base over from Flow (which is how I found this library to begin with). So I appreciate your help and patience here.

I didn't update the docs in this PR yet. I was planning to do that in a separate PR and include further proofreading changes. The next step will be figuring out `createAction` but that can be a separate PR, if you prefer. The intended distinction between `createAction` and `createStandardAction` is still a bit fuzzy to me.

File: src/action.spec.snap.ts [hunk 1]
```diff
@@ -82,3 +81,57 @@ describe('with optional payload and optional meta', () => {
...
+  ) => action('ACTION', message, scope, error);
+  // @dts-jest:pass:snap -> { type: "ACTION"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }
+  withOptionalPayloadOptionalMetaAndOptionalError(); // => { type: 'ACTION' }
+  // @dts-jest:pass:snap -> { type: "ACTION"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }
+  withOptionalPayloadOptionalMetaAndOptionalError('Hello!'); // => { type: 'ACTION', payload: 'Hello!' }
+  // @dts-jest:pass:snap -> { type: "ACTION"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }
+  withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info'); // => { type: 'ACTION', payload: 'Hello!', meta: 'info' }
+  // @dts-jest:pass:snap -> { type: "ACTION"; payload: string | undefined; meta: string | undefined; error: boolean | undefined; }
+  withOptionalPayloadOptionalMetaAndOptionalError('Hello!', 'info', true); // => { type: 'ACTION', payload: 'Hello!', meta: 'info', error: true }
+});
```
Comment (piotrwitek) [inline src/action.spec.snap.ts]: tests looking good!

File: src/action.spec.ts [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import * as T from './type-helpers';
```
Comment (piotrwitek) [inline src/action.spec.ts]: Please don't remove this one, it's important for CI, you can add comment or something: type-tests global

File: src/action.ts [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+): { type: T; error: E };
+
+export function action<T extends StringType, P, E>(
+  type: T,
+  payload: P,
+  meta: undefined,
+  error: E
+): { type: T; payload: P; error: E };
+
+export function action<T extends StringType, M, E>(
```
Comment (piotrwitek) [inline src/action.ts]: should be one position higher, because more specific arg is eariler

**Comment by piotrwitek** [comment]
@sjsyrek After few review comments should be good to merge

**Comment by piotrwitek** [comment]
> The intended distinction between `createAction` and `createStandardAction` is still a bit fuzzy to me.
Good that you're mentioning that. The core distinction is really the preference of the user to use generic type arguments (similar to `redux-actions`) or more inference driven approach, where types are inferred solely from function signature (args and returned object)

**Comment by sjsyrek** [comment]
@piotrwitek Should be good now

**Comment by piotrwitek** [comment]
Thanks @sjsyrek for adding a great new feature!

**Pull request merged**